### PR TITLE
Asset ViewHelper

### LINKED ISSUE #177: How to refresh browser cache after Deployment?
Sorry, but I can not understand, how can I update a browser cache after changes?
Is it possible at all in AssetManager?
I know two possibilities.
1. To set header Cache-control & Expires.
2. Or append some unique string to assets, ex. /assets/css/main.css?123
Also I have an idea, to append a last modified date, which I can get from asset file automatically and convert it to timestamp, ex. /assets/css/main.css?1231434698590
Can you help me please? Thx

### PR DESCRIPTION
**Pull request by pensiero**
The `Asset` view helper give the possibility refresh browser cache when assets files change.
`Asset` append a query param to the filename passed during the invocation of the view helper.

The query param appended is `u=[UNIX_TIMESTAMP]` where `[UNIX_TIMESTAMP]` is the last modification date of the file passed.

This PR is not completed and I need help:

1. Is it ok to use a [factory](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/Module.php#L91-L95) in the `Module`?
2. This helper works only if the file is written in the fs, so I need the `dir` option specified in the `asset_manager/caching` config array... now it's [hardcoded](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/View/Helper/Asset.php#L18-L20), but I think that there is a better way.. or it's ok this way?
3. I wrote a [test](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/tests/AssetManagerTest/View/Helper/AssetTest.php), obviously it fails.. any suggest is really well accepted to know how to procede.

Thanks everybody!

Fixes #177

### CODE REVIEW

File: src/AssetManager/Module.php [hunk 1]
```diff
@@ -83,6 +85,17 @@ public function onBootstrap(EventInterface $event)
         $eventManager->attach(MvcEvent::EVENT_DISPATCH_ERROR, $callback, $priority);
     }
 
+    public function getViewHelperConfig()
+    {
+        return [
```
Comment (RWOverdijk) [inline src/AssetManager/Module.php]: Long array notation

File: src/AssetManager/View/Helper/Asset.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            // find the file and if it exists, append its unix modification time to the filename
+            $originalPath = $assetsPath . $filename;
+            if (file_exists($originalPath)) {
+                return $filename . '?u=' . filemtime($originalPath);
+            }
+        }
+
+        return $filename;
+    }
+}
```
Comment (RWOverdijk) [inline src/AssetManager/View/Helper/Asset.php]: Newline

File: tests/AssetManagerTest/View/Helper/AssetTest.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+{
+    public function testInvoke()
+    {
+        $filename = '';
+        $config = [];
+        $helper = new Asset($config);
+
+        $this->assertContains("?u=", $helper->__invoke($filename));
+    }
+}
```
Comment (RWOverdijk) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: Newline

**Comment by RWOverdijk** [comment]
1. I think a factory is fine.
2. I'd say, make it configurable. All this does is load the file and make a hash. However, I would perhaps cache the hash based on file name to make sure you don't get a ton of stat calls per request.
3. I haven't looked at the test yet, will do soon.

Also:
- I'd make the query string parameter configurable, and default to `_`.

File: src/AssetManager/Module.php (hunk 1 above)
Comment (pensiero) [inline src/AssetManager/Module.php]: Ok

File: src/AssetManager/View/Helper/Asset.php (hunk 2 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: Ok

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 3 above)
Comment (pensiero) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: Ok

**Comment by pensiero** [comment]
1. Great
2. Do the user have to specify what cache system should be used?

For the configs, is it ok if I do something like I did? 
For example `if (isset($this->config['asset_manager']['caching']['default']['options']['dir'])) {`

THX

**Comment by RWOverdijk** [comment]
@pensiero I think that'd be wise. Any (@o

**Comment by pensiero** [comment]
Thanks @RWOverdijk, with the last commit the `AssetCacheManager` is not available in the newly created view helper, but now I don't know exactly how to get the cache from the AssetCacheManager and use to cache the file timestamp. Can you or @Ocramius help me?

Thanks a lot

**Comment by RWOverdijk** [comment]
@pensiero I'm not sure what you mean. I believe that's just DI. Could you also move the factory to it's own file?

**Comment by pensiero** [comment]
With the latest update you can specify a custom cache for the view helper (must be a Zend Cache Adapter).

If you do not specify anything in the config, no cache will be used.
A custom config should look like this:
```
'service_manager' => array(
    'factories'          => array(
        'Application\Cache\Redis' => 'Application\Cache\RedisFactory',
    ),
),
'asset_manager'   => array(
    'view_helper' => array(
        'query_string' => '_',
        'cache'        => 'Application\Cache\Redis',
    ),
    'caching'          => array(
        'default' => array(
            'cache'     => 'AssetManager\Cache\FilePathCache',
            'options' => array(
                'dir' => 'public/assets',
            ),
        ),
    ),
),
```

where the RedisFactory could be something like this:
```
<?php
namespace Application\Cache;

use Zend\ServiceManager\FactoryInterface;
use Zend\ServiceManager\ServiceLocatorInterface;
use Zend\Cache\Storage\Adapter\RedisOptions;
use Zend\Cache\Storage\Adapter\Redis;

class RedisFactory implements FactoryInterface
{
    const DEFAULT_RESULT_CACHE_TTL = 300;

    public function createService(ServiceLocatorInterface $serviceLocator)
    {
        $redisOptions = new RedisOptions();
        $redisOptions
            ->setServer(array('127.0.0.1', '6379'))
            ->setNamespace('Application_Redis')
            ->setPersistentId('application_redis')
            ->setTtl(self::DEFAULT_RESULT_CACHE_TTL);

        $redisOptions->setLibOptions([
            \Redis::OPT_SERIALIZER => \Redis::SERIALIZER_PHP
        ]);

        $redis = new Redis($redisOptions);

        return $redis;
    }
}
```

The implementation of the cache is really easy and friendly to mantain. I've also improved and optimizated the code.

Finally, I update also something in the test file, but I'm not able to complete.. can someone help me with that?
I think it could be enough to:
- Invoke without the cache and assert that the output contain something like `?_=NUMBER`
- Invoke with the cache and assert that the output contain something like `?_=NUMBER`


Thanks all

**Comment by pensiero** [comment]
@RWOverdijk @Ocramius there is only one problem: how can I check the `filemtime` only AFTER the creation of the file? Because it seems that the file is created after the request to the view helper, and so the helper can't find it.

**Comment by RWOverdijk** [comment]
@pensiero A cache indeed looks something like that.

As to the cache, [it's being set here](https://github.com/RWOverdijk/AssetManager/blob/fed0e5cc6799a595627c79c41da4492f6e857cc0/src/AssetManager/Service/AssetManager.php#L146).

**Comment by pensiero** [comment]
So it's ok how I did? And for the problem of how the check after the creation of the file? Thanks

**Comment by RWOverdijk** [comment]
@pensiero As to the problem, you use resolve to find the asset for the ViewHelper, so you also set the asset on the cache. This would be duplicating logic (setting the cache in the manager and the viewhelper) so perhaps move that to a method. I'm curious what @Ocramius thinks of all of this.

**Comment by pensiero** [comment]
I was thinking about using the same cache used by the file, but think about if we set (as in this case) a Filesystem cache:
- we cache the result of a `filemtime` request, writing it in a file
- then we search for this file and we read its content

In this case it's really stupid to do that... rather its better to call `filemtime` everytime the ViewHelper is invoked.

So my solution is to use a more fast and flexible cache system, using one Adapter provided by Zend Framework.

**Comment by pensiero** [comment]
Any news about it? THX

**Comment by RWOverdijk** [comment]
@pensiero I'm not sure if I understand what you're saying... But I'd use the path of the requested file as a key, and the time as the value and put that in a cache. Looking it up every time is a bad idea because... Well, performance.

**Comment by pensiero** [comment]
@RWOverdijk Yes, in fact with the [latest commit](https://github.com/pensiero/AssetManager/commit/504a48b15ab8356a925503ffc54c6079b7b10c17) is it possible to specify a custom cache adapter to cache the result of the `filemtime` method.

The only problem is that the `filemtime` method is invoked before the creation of the file, so the first time it's invoked, the result is null because it doesn't find the file.

**Comment by RWOverdijk** [comment]
@pensiero So perhaps you should call resolve() from the view helper to force it to cache?

File: src/AssetManager/Module.php [hunk 4]
```diff
@@ -83,6 +86,17 @@ public function onBootstrap(EventInterface $event)
         $eventManager->attach(MvcEvent::EVENT_DISPATCH_ERROR, $callback, $priority);
     }
 
+    public function getViewHelperConfig()
+    {
+        return array(
+            'factories' => array(
```
Comment (RWOverdijk) [inline src/AssetManager/Module.php]: Please move to a Factory class

File: src/AssetManager/View/Helper/Asset.php [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    private $serviceLocator;
+
+    /**
+     * @param ServiceLocatorInterface $serviceLocator
+     */
+    public function __construct(ServiceLocatorInterface $serviceLocator)
+    {
+        $this->serviceLocator = $serviceLocator;
+        $this->config = $serviceLocator->get('config')['asset_manager'];
```
Comment (RWOverdijk) [inline src/AssetManager/View/Helper/Asset.php]: Align consecutive assignments

**Comment by pensiero** [comment]
@RWOverdijk i never use that method, how can I use it and where? in the view or inside the view helper? thanks a lot

File: src/AssetManager/View/Helper/Asset.php (hunk 5 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: :+1:

File: src/AssetManager/Module.php (hunk 4 above)
Comment (pensiero) [inline src/AssetManager/Module.php]: :+1:

**Comment by RWOverdijk** [comment]
@pensiero In the view helper. :) The method is here: https://github.com/RWOverdijk/AssetManager/blob/master/src/AssetManager/Service/AssetManager.php#L185
The cache manager is being set here: https://github.com/RWOverdijk/AssetManager/blob/fed0e5cc6799a595627c79c41da4492f6e857cc0/src/AssetManager/Service/AssetManagerServiceFactory.php#L40,L42 And being cached here: https://github.com/RWOverdijk/AssetManager/blob/fed0e5cc6799a595627c79c41da4492f6e857cc0/src/AssetManager/Service/AssetManager.php#L146

So you could do a lookup in the viewhelper (you'll have to anyway, to make sure you have the right asset) and then cache it. Ensuring the file exists, you can create all kinds of silly magic to map the file (resolved file path!) to a timestamp.

**Comment by pensiero** [comment]
@RWOverdijk i tried with:
```
$asset = $this->serviceLocator->get('AssetManager\Service\AssetManager')->getResolver()->resolve($filename);
```

in the view helper, but it returns null.
`$filename` is what I pass to the view helper (for example, `/js/main.js`)

**Comment by RWOverdijk** [comment]
@pensiero So maybe it can't find the file. I'm currently not able to check this for you sorry. But the AssetManager [does some stuff](https://github.com/RWOverdijk/AssetManager/blob/master/src/AssetManager/Service/AssetManager.php#L191,L197). Perhaps see what you're skipping?

**Comment by pensiero** [comment]
SOOOLVED!

Using the [getLastModified()](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/Asset/AggregateAsset.php#L62) method instead of the `filemtime`, and [resolving the Asset before](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/View/Helper/Asset.php#L41), everything work!

I also modified the code according to your suggestions, @RWOverdijk.

Thanks a lot :)

**Comment by RWOverdijk** [comment]
Good to hear :)

**Comment by pensiero** [comment]
:+1: when you are ready, you can merge it ;)

File: src/AssetManager/Module.php (hunk 4 above)
Comment (RWOverdijk) [inline src/AssetManager/Module.php]: Also move the factory to the config with the other factories. Sorry. I thought you knew :)

File: src/AssetManager/View/Helper/Asset.php [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    public function __invoke($filename)
+    {
+        // search the cache config for the specific file requested (if none, use the default one)
+        if (isset($this->config['caching'][$filename])) {
+            $cacheConfig = $this->config['caching'][$filename];
+        } else if (isset($this->config['caching']['default'])) {
+            $cacheConfig = $this->config['caching']['default'];
+        }
+
+        if (isset($cacheConfig) && isset ($cacheConfig['options']['dir'])) {
```
Comment (RWOverdijk) [inline src/AssetManager/View/Helper/Asset.php]: 2 remarks.

1. `isset` allows nested values to be supplied. This means that `isset($cacheConfig['options']['dir'])` also works
2. Return early. `if (!isset($cacheConfig['options']['dir'])) { return $filename; }`. Less indentation is happiness

File: tests/AssetManagerTest/View/Helper/AssetTest.php [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+use PHPUnit_Framework_TestCase as TestCase;
+use AssetManager\View\Helper\Asset;
+use Zend\ServiceManager\ServiceManager;
+
+class AssetTest extends TestCase
+{
+    public function testInvoke()
+    {
+        /* TODO need help
```
Comment (RWOverdijk) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: Need help?

**Comment by RWOverdijk** [comment]
@Ocramius Could you also cast an eye on this PR?

File: src/AssetManager/View/Helper/Asset.php (hunk 6 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: :+1:

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 7 above)
Comment (pensiero) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: Yes sir, I'm not good with tests :(

File: src/AssetManager/Module.php (hunk 4 above)
Comment (pensiero) [inline src/AssetManager/Module.php]: :+1:

**Comment by pensiero** [comment]
Any news?

**Comment by RWOverdijk** [comment]
I think my ping to @Ocramius got buried in the pile of emails he gets daily. Let's see if this ping to @Ocramius does get through. I still think the test should be fixed. What is it that's making it difficult for you?

**Comment by pensiero** [comment]
I don't know how to invoke the ServiceManager needed by the view Helper

**Comment by Ocramius** [comment]
Ping got through, but no time ATM

File: src/AssetManager/Service/AssetViewHelperFactory.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
+<?php
+namespace AssetManager\Service;
+
+use Zend\ServiceManager\FactoryInterface;
+use Zend\ServiceManager\ServiceLocatorInterface;
+use AssetManager\View\Helper\Asset;
+
+class AssetViewHelperFactory implements FactoryInterface {
```
Comment (Ocramius) [inline src/AssetManager/Service/AssetViewHelperFactory.php]: Note: you should respect PSR-2 ;-)

File: src/AssetManager/View/Helper/Asset.php [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+    /**
+     * @var ServiceLocatorInterface
+     */
+    private $serviceLocator;
+
+    /**
+     * @param ServiceLocatorInterface $serviceLocator
+     */
+    public function __construct(ServiceLocatorInterface $serviceLocator)
```
Comment (Ocramius) [inline src/AssetManager/View/Helper/Asset.php]: Don't inject the service locator. Inject the config from the asset manager, as well as the asset manager itself.

File: src/AssetManager/View/Helper/Asset.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+     *
+     * @param string $assetsPath
+     * @param string $filename
+     * @param string $queryString
+     * @return string
+     */
+    private function elaborateFilePath($assetsPath, $filename, $queryString)
+    {
+        // resolve asset
+        $asset = $this->serviceLocator->get('AssetManager\Service\AssetManager')->getResolver()->resolve($filename);
```
Comment (Ocramius) [inline src/AssetManager/View/Helper/Asset.php]: Same as above

File: src/AssetManager/View/Helper/Asset.php [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+        if ($asset !== null) {
+
+            // append last modified date to the filepath and use a custom query string
+            return $filename . '?' . $queryString . '=' . $asset->getLastModified();
```
Comment (Ocramius) [inline src/AssetManager/View/Helper/Asset.php]: Should it be encoded?

File: src/AssetManager/View/Helper/Asset.php [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * Output the filepath with its unix modification time as query param
+     *
+     * @param string $filename
+     * @return string
+     */
+    public function __invoke($filename)
+    {
+        // search the cache config for the specific file requested (if none, use the default one)
+        if (isset($this->config['caching'][$filename])) {
```
Comment (Ocramius) [inline src/AssetManager/View/Helper/Asset.php]: This method NEEDS splitting.

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 7 above)
Comment (Ocramius) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: @pensiero you need to test the helper, sorry. Feel free to poke me on hangouts about that, although I will be busy until Oct 25.

File: src/AssetManager/Service/AssetViewHelperFactory.php (hunk 8 above)
Comment (pensiero) [inline src/AssetManager/Service/AssetViewHelperFactory.php]: :+1:

File: src/AssetManager/View/Helper/Asset.php (hunk 9 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: I can pass the config, but I need the serviceLocator not only for getting the AssetManager, but also for the cache [on line 85 of this file](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/View/Helper/Asset.php#L85)

File: src/AssetManager/View/Helper/Asset.php (hunk 11 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: Everything or only the `$queryString`? I think that is only necessary to encode the `$queryString`

File: src/AssetManager/View/Helper/Asset.php (hunk 12 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: :+1:

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 7 above)
Comment (pensiero) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: How can I instantiate a ServiceManager (comprehensive all configs) to pass to the view helper?

**Comment by pensiero** [comment]
Last commit solve something reported by @Ocramius 
It remains only two things:
- [ ] Decide if to move [this lines](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/View/Helper/Asset.php#L70-L91) into the [AssetViewHelperFactory](https://github.com/pensiero/AssetManager/blob/feature/asset-view-helper/src/AssetManager/Service/AssetViewHelperFactory.php)
- [ ] Know how to instantiate a ServiceManager (with configs) to pass to the Asset ViewHelper for testing purposes

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 7 above)
Comment (Ocramius) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: You don't need the `ServiceManager` if you do proper injection and mock all the dependencies.

File: src/AssetManager/View/Helper/Asset.php [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+     * @var ServiceLocatorInterface
+     */
+    private $assetManager;
+
+    /**
+     * @param ServiceLocatorInterface $serviceLocator
+     * @param AssetManager            $assetManager
+     * @param array                   $config
+     */
+    public function __construct(ServiceLocatorInterface $serviceLocator, AssetManager $assetManager, $config)
```
Comment (Ocramius) [inline src/AssetManager/View/Helper/Asset.php]: Don't inject the `ServiceLocatorInterface`.

File: src/AssetManager/View/Helper/Asset.php [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+     * find the file and if it exists, append its unix modification time to the filename
+     *
+     * @param string $filename
+     * @param string $queryString
+     * @return string
+     */
+    private function elaborateFilePath($filename, $queryString)
+    {
+        /** @var AggregateAsset $asset */
+        $asset = $this->serviceLocator->get('AssetManager\Service\AssetManager')->getResolver()->resolve($filename);
```
Comment (Ocramius) [inline src/AssetManager/View/Helper/Asset.php]: Don't call `->getResolver()` on the `AssetManager`. Inject it directly instead (constructor)

File: src/AssetManager/View/Helper/Asset.php [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        // check if the cache is configured
+        if (!isset($this->config['view_helper']['cache']) || $this->config['view_helper']['cache'] == null) {
+            return null;
+        }
+
+        // get the cache, if it's a string, search it among services
+        $cache = $this->config['view_helper']['cache'];
+        if (is_string($cache)) {
+            $cache = $this->serviceLocator->get($cache);
```
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: @Ocramius if I do not inject the serviceLocator, how can I get the cache here?

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 7 above)
Comment (pensiero) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: Can you provide me some examples on other projects? THX

File: src/AssetManager/View/Helper/Asset.php (hunk 15 above)
Comment (RWOverdijk) [inline src/AssetManager/View/Helper/Asset.php]: You can specify it as a dependency :)

Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: Do you have a similar working piece of code as example?

Comment (RWOverdijk) [inline src/AssetManager/View/Helper/Asset.php]: https://github.com/RWOverdijk/AssetManager/pull/180/files#diff-25c0bbb35fb4899efece895d4b61b011R27

The factory does exactly that, inject dependencies. That's where you should get them :)

https://github.com/RWOverdijk/AssetManager/blob/master/src/AssetManager/Service/AssetFilterManagerServiceFactory.php#L27

Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: :+1: done, what do you think about it?

File: src/AssetManager/View/Helper/Asset.php [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+     */
+    private $assetManagerResolver;
+
+    /**
+     * @param ServiceLocatorInterface   $serviceLocator
+     * @param ResolverInterface         $assetManagerResolver
+     * @param AbstractCacheAdapter|null $cache
+     * @param array                     $config
+     */
+    public function __construct(ServiceLocatorInterface $serviceLocator, ResolverInterface $assetManagerResolver, AbstractCacheAdapter $cache, $config)
```
Comment (RWOverdijk) [inline src/AssetManager/View/Helper/Asset.php]: You can remove the locator now

**Comment by RWOverdijk** [comment]
Looks better. Some feedback and broken tests, but other than that I'm starting to like it! I know a lot of work has gone into this PR so far, thank you for that.

File: src/AssetManager/View/Helper/Asset.php (hunk 16 above)
Comment (pensiero) [inline src/AssetManager/View/Helper/Asset.php]: you're right!

**Comment by pensiero** [comment]
@RWOverdijk it's a pleasure for me and a great occasion to learn :)

File: tests/AssetManagerTest/View/Helper/AssetTest.php (hunk 7 above)
Comment (RWOverdijk) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: Sure can, it's being used in other parts of the test suite, too. :)

Comment (pensiero) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: @RWOverdijk where exactly?

Comment (RWOverdijk) [inline tests/AssetManagerTest/View/Helper/AssetTest.php]: https://github.com/RWOverdijk/AssetManager/blob/master/tests/AssetManagerTest/Service/AssetManagerTest.php#L45,L58

Here you can see it constructs its own MimeResolver, and then mocks a resolver. Does this help?

**Comment by RWOverdijk** [comment]
@pensiero I've been a bit quiet lately, sorry for that. How's it going with this?

**Comment by pensiero** [comment]
@RWOverdijk sorry, really in troubles during last months, but it's not an excuse, I have to find some time to see this

**Comment by RWOverdijk** [comment]
It's a fine excuse :) Good luck with your trouble!

**Comment by pensiero** [comment]
@RWOverdijk can you take a look? Tests should be good :)

File: src/AssetManager/Service/AssetViewHelperFactory.php [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    /**
+     * @param ServiceLocatorInterface $serviceLocator
+     * @param array                   $config
+     *
+     * @return null
+     */
+    private function loadCache($serviceLocator, $config)
+    {
+        // check if the cache is configured
+        if (!isset($config['view_helper']['cache']) || $config['view_helper']['cache'] == null) {
```
Comment (RWOverdijk) [inline src/AssetManager/Service/AssetViewHelperFactory.php]: ===

**Comment by RWOverdijk** [comment]
@Ocramius I think I'm ready to merge this in. I just need you to agree / disagree so we can move on :)

**Comment by RWOverdijk** [comment]
@Ocramius fyi merging this tonight.

**Pull request merged**